### PR TITLE
Fleet + snapshot & restore tests

### PR DESCRIPTION
**Pull request by slickwarren**
## Issue: 

https://github.com/rancher/qa-tasks/issues/1525
 
## Problem

adding fleet tests where snapshots are taken in the process
 
## Solution

* move etcd snapshot helpers -> actions
* add folder for snapshots in fleet (validation/fleet/snapshot)
 
## Testing

tested on rke1 and rke2 for snapshot only
tested on rke2 for fleet + snapshots (rke1 not supported due to cluster checks in setup, can add later if need be)

## QA Testing Considerations

 https://github.com/rancher/shepherd/pull/304
see notes in ^ for some improvements in shepherd and some other considerations

### CODE REVIEW

File: tests/v2/validation/provisioning/rke1/custom_cluster_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/custom_cluster_test.go:76]: These are commented out when they shouldn't be.

File: tests/v2/validation/provisioning/rke1/custom_cluster_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/custom_cluster_test.go:77]: Same here.

File: tests/v2/validation/provisioning/rke1/custom_cluster_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/custom_cluster_test.go:88]: Same as above.

File: tests/v2/validation/provisioning/rke1/custom_cluster_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/custom_cluster_test.go:89]: Same.

File: tests/v2/validation/provisioning/rke1/provisioning_node_driver_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/provisioning_node_driver_test.go:76]: Same as above.

File: tests/v2/validation/provisioning/rke1/provisioning_node_driver_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/provisioning_node_driver_test.go:77]: Same as above.

File: tests/v2/actions/etcdsnapshot/etcdsnapshot.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/etcdsnapshot.go:137]: I'm curious if this has been tested for S3? This specifically doesn't yield the exact results as local snapshots, so just calling it out.

File: tests/v2/actions/etcdsnapshot/verify.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/verify.go:59]: I think having only the return statement is fine here, but not necessarily opposed if you want to keep the logs.

**Review by markusewalker — COMMENTED** [review]
Left some comments. This seems to be touching quite a bit of `snapshot` package that isn't just fleet specific. It would be good to have testing for local and S3 for each cluster type.

File: tests/v2/actions/etcdsnapshot/etcdsnapshot.go
Comment (slickwarren) [reply tests/v2/actions/etcdsnapshot/etcdsnapshot.go:137]: ack

File: tests/v2/actions/etcdsnapshot/verify.go
Comment (slickwarren) [reply tests/v2/actions/etcdsnapshot/verify.go:67]: ack

File: tests/v2/validation/fleet/snapshot/README.md
Comment (anupama2501) [inline tests/v2/validation/fleet/snapshot/README.md:25]: aren't these values optional or they need to be set by the user is it?

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:107]: I don't think saving these for the tests is what we should do in this case. We don't want people using the same values for this every time. I feel setting a global const will make people think they should be using those instead of setting their own.

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:95]: ack

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:97]: ack

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:32]: ack

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:156]: Deployments are part of the tests for snapshots. I didn't think it would make sense to not validate everything we normally do for snapshots. Rather I wanted to validate everything we do for snapshots and in addition, add fleet into the mix.

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:165]: ack

I can add rke1 support later. tldr, I didn't support it in other fleet packages and that was causing problems here.

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:173]: those are included in the called functions themselves, so that the test isn't cluttered with logs.

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:185]: ack

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:231]: name of the test is modified, so I wasn't sure how to do that well if this lives in setup.

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (slickwarren) [reply tests/v2/actions/etcdsnapshot/creates.go:38]: this was mostly moving stuff over, I didn't feel it appropriate to change the existing consts. But its usually better to have a const than multiple of the same string in your code.

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (slickwarren) [reply tests/v2/actions/etcdsnapshot/creates.go:62]: ack. But I don't think refactoring is necessary, really. This is a helper function to do both creating and validation, which should still fit our rules for being an action. I'll wait for Hostbusters input for a second opinion

File: tests/v2/validation/fleet/snapshot/README.md
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/README.md:25]: I don't have a dynamic test so this can be removed, as I hard coded values to make it simpler to run. I don't see value in having a dynamic test for this use case.

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (lscalabrini01) [inline tests/v2/actions/etcdsnapshot/creates.go:530]: Could you add the function name on the godoc?

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (lscalabrini01) [inline tests/v2/actions/etcdsnapshot/creates.go:574]: Same here

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (lscalabrini01) [inline tests/v2/actions/etcdsnapshot/creates.go:701]: And here

File: tests/v2/actions/etcdsnapshot/verify.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/verify.go:20]: Since you have these broken up in two lines, can you start the second line with `and the correct number...`? It looks a bit awkward currently with how it's broken up.

File: tests/v2/actions/scalinginput/replace.go
Comment (markusewalker) [inline tests/v2/actions/scalinginput/replace.go:40]: We should probably add a godoc here, I didn't realize we didn't have one for this function.

File: tests/v2/validation/provisioning/rke2/custom_cluster_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke2/custom_cluster_test.go:155]: This should be uncommented.

File: tests/v2/validation/provisioning/rke2/provisioning_node_driver_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke2/provisioning_node_driver_test.go:75]: All of these need to be uncommented out.

File: tests/v2/validation/provisioning/rke1/provisioning_node_driver_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke1/provisioning_node_driver_test.go:79]: Why are we changing it to this? The static tests are testing the various cluster configurations. If anything, we should probably add 3, 2, 3 as a static option, but not the only option.

File: tests/v2/validation/provisioning/rke2/provisioning_node_driver_test.go
Comment (markusewalker) [inline tests/v2/validation/provisioning/rke2/provisioning_node_driver_test.go:79]: Why are we changing it to this? The static tests are testing the various cluster configurations. If anything, we should probably add 3, 2, 3 as a static option, but not the only option.

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/creates.go:61]: Can you omit everything after `original snapshot`? Also, can you just have the comment as two lines to make it look a bit cleaner?

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/creates.go:92]: Can you remove this extra line?

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/creates.go:187]: This is interesting, so in essence, if S3 is enabled but a local snapshot is selected, the test is going to fail here? Is this something we really want to be doing? It almost seems that if this is the case, we should enhance the logic which I think you already did.

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (markusewalker) [inline tests/v2/actions/etcdsnapshot/creates.go:278]: Can you have this third comment on the same line as the second comment?

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (anupama2501) [reply tests/v2/actions/etcdsnapshot/creates.go:38]: I understand the importance of having constants in most cases, but in this instance, I don’t think defining all = "all" or port = "port" adds much value. These values are self-explanatory, and unlikely to change. Keeping them as string literals avoids unnecessary complexity and improves readability. but this need not be updated in this pr as this is just porting changes.

File: tests/v2/validation/provisioning/rke2/provisioning_node_driver_test.go
Comment (slickwarren) [reply tests/v2/validation/provisioning/rke2/provisioning_node_driver_test.go:79]: sorry this was for testing, I'll remove it.

**Review by markusewalker — DISMISSED** [review]

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (anupama2501) [reply tests/v2/actions/etcdsnapshot/creates.go:61]: Agreed refactoring shouldn’t happen in this PR to avoid the risk of breaking existing tests, with how involved the snapshot suite is. but, I think it would be beneficial to create a ticket to revisit this later. We can discuss the potential risks and benefits of separating the Create and Validate logic in our design call, to make sure the tests are easily maintainable.

File: tests/v2/validation/fleet/provisioning/new_cluster_existing_gitrepo_test.go
Comment (anupama2501) [inline tests/v2/validation/fleet/provisioning/new_cluster_existing_gitrepo_test.go:124]: was just noticing this. qq here. are we using hardened cluster in the tests?

File: tests/v2/validation/fleet/public_gitrepo_test.go
Comment (anupama2501) [inline tests/v2/validation/fleet/public_gitrepo_test.go:144]: Noticed we have same log comments but different parameters. is this intentional. if yes, can we combine both in one statement?

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (anupama2501) [inline tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:107]: I think it would be more appropriate to call this description instead of name based on line 110 :)

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (anupama2501) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go]: I’m a bit confused on why deployments are a part of the snapshot tests. The primary goal of snapshot is to make sure that the resources persist before and after the snapshot, for which we're specifically validating the Fleet resources. In my opinion, including deployments isn't necessary since the tests are mainly to verify fleet resources are intact before and after a snapshot.

**Review by anupama2501 — COMMENTED** [review]
few nits and comments.

File: tests/v2/actions/etcdsnapshot/creates.go
Comment (slickwarren) [reply tests/v2/actions/etcdsnapshot/creates.go:61]: https://github.com/rancher/qa-tasks/issues/1582

File: tests/v2/validation/fleet/provisioning/new_cluster_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/provisioning/new_cluster_existing_gitrepo_test.go:124]: This was an update to provision hardened cluster instead of non-hardened. Since hardened is a subset of normal provisioning, it makes the most sense to just change this to hardened vs. having 2 tests for it.

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go:107]: this is 'name of test' which follows the same name in basically all the tests we have.

File: tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go
Comment (slickwarren) [reply tests/v2/validation/fleet/snapshot/snapshot_restore_existing_gitrepo_test.go]: we want to append to the existing snapshot test. The test is snapshot + fleet, so why would I change what is tested as part of snapshots outside of _adding_ tests for fleet?

**Review by anupama2501 — APPROVED** [review]
approving but as note: go.mod and sum needs to be changed before its ready to be merged.

**Review by markusewalker — APPROVED** [review]

**Review by markusewalker — APPROVED** [review]

**Review by anupama2501 — APPROVED** [review]

**Pull request merged**
